pyanaconda/modules/payloads/payload/dnf/utils.py

```diff
@@ -184,3 +187,103 @@ def get_kernel_version_list():
+    """
+    mapping = {}
+
+    # Generate the dictionary of mount points and sizes.
+    output = execWithCapture('df', ['--output=target,avail'])
+    lines = output.rstrip().splitlines()
+
+    for line in lines:
+        key, val = line.rsplit(maxsplit=1)
+
+        if not key.startswith('/'):
+            continue
+
+        mapping[key] = Size(int(val) * 1024)
+
+    # Add /var/tmp/ if this is a directory or image installation.
+    if not conf.target.is_hardware:
+        var_tmp = os.statvfs("/var/tmp")
+        mapping["/var/tmp"] = Size(var_tmp.f_frsize * var_tmp.f_bfree)
+
+    return mapping
+
+
+def pick_mount_point(mount_points, download_size, install_size, download_only=False):
+    """Pick a mount point for the package installation.
+
+    :param mount_points: a dictionary of mount points and their available spaces
+    :param download_size: a total size required for the package download
+    :param install_size: a total size required for the package installation
+    :param download_only: pick a mount point only for the package download
+    :return: a mount point or None
+    """
+    log.info("Pick a mount point for %s download, %s install, %s total.",
+             download_size, install_size, download_size + install_size)
+
+    # Find sufficient mount points.
+    sufficient = _pick_mount_points(
+        mount_points,
+        download_size,
+        install_size
+    )
+
+    # Or find sufficient mount points specifically for the download.
+    if download_only:
+
+        # Ignore the installation size.
+        if not sufficient:
+            sufficient = _pick_mount_points(
+                mount_points,
+                download_size,
+                install_size=0
+            )
+
+        # Ignore the system root if there are other mount points.
+        if len(sufficient) > 1:
+            sufficient.discard(conf.target.system_root)
+
+    # Choose the biggest sufficient mount point.
+    mount_point = max(sufficient, default=None, key=mount_points.get)
+
+    log.info("Pick a sufficient mount point: %s", mount_point)
+    return mount_point
+
+
+def _pick_mount_points(mount_points, download_size, install_size):
+    """Pick mount points for the package installation.
+
+    :return: a set of sufficient mount points
+    """
+    log.debug("Mount points to choose from: %s", mount_points)
+    sufficient = set()
+
+    suitable = {
+        '/var/tmp',
+        conf.target.system_root,
+        join_paths(conf.target.system_root, 'home'),
+        join_paths(conf.target.system_root, 'tmp'),
+        join_paths(conf.target.system_root, 'var'),
+    }
+
+    for mount_point, size in mount_points.items():
+
+        # Ignore mount points that are not suitable.
+        if mount_point not in suitable:
+            continue
+
+        elif size > (download_size + install_size):
+            sufficient.add(mount_point)
+
+        elif mount_point != conf.target.system_root and size > download_size:
```

Comment:
Shouldn't we check if `mount_point` starts with system root?

Reply:
This is the original implementation, but I had the same thought. I can fix it in another commit.

Reply:
Fixed.